Windows support for sdist build

TWAC opened the pull request:
Not for merge, just for checking buildbots

TWAC commented:
If the buildbots pass, please check if this is mergeable.

MathMagique commented:
Will check.

MathMagique commented on cpp/cpp_odbc/Library/cpp_odbc/multi_value_buffer.h:
```diff
@@ -1,17 +1,18 @@
 struct writable_buffer_element {
 	char * data_pointer;
-	long & indicator;
+	intptr_t & indicator;
```
I guess this should be `int64_t`, as the indicator is always an `SQLLEN` value and is not associated to any pointer type at all.

MathMagique replied:
this comment applies for the rest of the file and the unit tests

MathMagique commented on cpp/cpp_odbc/Test/tests/multi_value_buffer_test.cpp:
```diff
@@ -80,7 +80,7 @@ TEST(MultiValueBufferTest, MutableElementAccess)
 	EXPECT_EQ( 0, std::memcmp(buffer.data_pointer(), "abcdef", 6));
 
-	long const expected_indicator = 42;
+	intptr_t const expected_indicator = 42;
```
See my comment on choosing `int64_t` over `intptr_t` for `multi_value_buffer`

MathMagique commented on cpp/turbodbc/Library/src/field_translators/int64_translator.cpp:
```diff
@@ -11,7 +11,7 @@ int64_translator::~int64_translator() = default;
 field int64_translator::do_make_field(char const * data_pointer) const
 {
-	return {*reinterpret_cast<long const *>(data_pointer)};
+	return {*reinterpret_cast<intptr_t const *>(data_pointer)};
```
This will translate to a 64 bit integer by design because the binding uses `SQL_BIGINT`. Please replace this with `int64_t`

MathMagique commented on cpp/turbodbc/Library/src/field_translators/timestamp_translator.cpp:
```diff
@@ -15,7 +15,7 @@ field timestamp_translator::do_make_field(char const * data_pointer) const
 	auto const ts = reinterpret_cast<SQL_TIMESTAMP_STRUCT const *>(data_pointer);
 	// map SQL nanosecond precision to posix_time microsecond precision
-	long const adjusted_fraction = ts->fraction / 1000;
+	intptr_t const adjusted_fraction = ts->fraction / 1000;
```
No need for `intptr_t` here; I'd prefer `int64_t`, then.

MathMagique commented on cpp/turbodbc/Library/src/make_description.cpp:
```diff
@@ -47,7 +47,7 @@ namespace {
 
 	struct description_by_value : public boost::static_visitor<description_ptr> {
-		description_ptr operator()(long const &) const
+		description_ptr operator()(intptr_t const &) const
```
Also, since this deals with a 64bit integer parameter, I'd prefer `int64_t` here.

MathMagique commented on cpp/turbodbc/Library/src/parameter_sets/set_field.cpp:
```diff
@@ -20,7 +20,7 @@ namespace {
 		}
 
-		bool operator()(long const &) const {
+		bool operator()(intptr_t const &) const {
```
`int64_t`

MathMagique commented on cpp/turbodbc/Library/src/parameter_sets/set_field.cpp:
```diff
@@ -57,10 +57,10 @@ namespace {
 		}
 
-		void operator()(long const & value)
+		void operator()(intptr_t const & value)
```
`int64_t`

MathMagique commented on cpp/turbodbc/Library/src/command.cpp:
```diff
@@ -55,7 +55,7 @@ bound_parameter_set & command::get_parameters()
 }
 
-long command::get_row_count()
+intptr_t command::get_row_count()
```
rowcount also returns SQLLEN, so it should return `int64_t` here.

MathMagique commented on cpp/turbodbc/Library/src/cursor.cpp:
```diff
@@ -67,7 +67,7 @@ std::shared_ptr<result_sets::result_set> cursor::get_result_set() const
 }
 
-long cursor::get_row_count()
+intptr_t cursor::get_row_count()
```
`int64_t` since row count returns `SQLLEN`

MathMagique commented on cpp/turbodbc/Library/src/descriptions/integer_description.cpp:
```diff
@@ -18,7 +18,7 @@ integer_description::~integer_description() = default;
 std::size_t integer_description::do_element_size() const
 {
-	return sizeof(long);
+	return sizeof(intptr_t);
```
`int64_t`

MathMagique commented on cpp/turbodbc/Library/turbodbc/command.h:
```diff
@@ -36,7 +36,7 @@ class command {
     bound_parameter_set &get_parameters();
 
-    long get_row_count();
+    intptr_t get_row_count();
```
`int64_t`

MathMagique commented on cpp/turbodbc/Library/turbodbc/cursor.h:
```diff
@@ -26,7 +26,7 @@ class cursor {
     void add_parameter_set(std::vector <nullable_field> const &parameter_set);
 
-    long get_row_count();
+    intptr_t get_row_count();
```
`int64_t`

MathMagique commented on cpp/turbodbc/Library/turbodbc/field.h:
```diff
@@ -12,7 +12,7 @@ namespace turbodbc {
  *        with a given row and column
  */
-using field = boost::variant<	long,
+using field = boost::variant<	intptr_t,
```
`int64_t` is the correct type since everywhere `SQL_BIGINT` is used for binding to retrieve 64 bit data.

MathMagique commented on cpp/turbodbc/Test/tests/command_test.cpp:
```diff
@@ -55,7 +55,7 @@ namespace {
 TEST(CommandTest, GetRowCountAfterQueryWithResultSet)
 {
-    long const expected = 17;
+    intptr_t const expected = 17;
```
`int64_t`

MathMagique commented on cpp/turbodbc/Test/tests/descriptions/integer_description_test.cpp:
```diff
@@ -8,7 +8,7 @@ TEST(IntegerDescriptionTest, BasicProperties)
 	turbodbc::integer_description const description;
 
-	EXPECT_EQ(sizeof(long), description.element_size());
+	EXPECT_EQ(sizeof(intptr_t), description.element_size());
```
`int64_t`

MathMagique commented on cpp/turbodbc/Test/tests/field_translators/int64_translator_test.cpp:
```diff
@@ -14,6 +14,6 @@ TEST(Int64TranslatorTest, MakeField)
 	int64_translator const translator;
 
-	*reinterpret_cast<long *>(element.data_pointer) = 42;
+	*reinterpret_cast<intptr_t *>(element.data_pointer) = 42;
```
`int64_t`

MathMagique commented on cpp/turbodbc/Test/tests/parameter_sets/set_field_test.cpp:
```diff
@@ -123,7 +123,7 @@ TEST(SetFieldTest, SetFieldInteger)
 
 	set_field(turbodbc::field{42l}, element);
-	EXPECT_EQ(42, *reinterpret_cast<long *>(element.data_pointer));
+	EXPECT_EQ(42, *reinterpret_cast<intptr_t *>(element.data_pointer));
```
`int64_t`

MathMagique commented on cpp/turbodbc/Test/tests/result_sets/double_buffered_result_set_test.cpp:
```diff
@@ -131,8 +131,8 @@ namespace {
 					auto element = (*buffer_)[i];
-					*reinterpret_cast<long *>(element.data_pointer) = (batch_index_ + 1);
-					element.indicator = sizeof(long);
+					*reinterpret_cast<intptr_t *>(element.data_pointer) = (batch_index_ + 1);
```
`int64_t`

MathMagique commented on cpp/turbodbc/Test/tests/result_sets/double_buffered_result_set_test.cpp:
```diff
@@ -171,17 +171,17 @@ TEST(DoubleBufferedResultSetTest, GetBuffers)
 	ASSERT_EQ(4, rs.fetch_next_batch());
 	ASSERT_EQ(1, rs.get_buffers().size());
-	EXPECT_EQ(1, *reinterpret_cast<long const *>(rs.get_buffers()[0].get()[3].data_pointer));
+	EXPECT_EQ(1, *reinterpret_cast<intptr_t const *>(rs.get_buffers()[0].get()[3].data_pointer));
```
`int64_t`

MathMagique commented on cpp/turbodbc/Test/tests/result_sets/field_result_set_test.cpp:
```diff
@@ -30,14 +30,14 @@ TEST(FieldResultSetTest, GetColumnInfoForwards)
 TEST(FieldResultSetTest, FetchRow)
 {
-	long const expected_int = 42;
+	intptr_t const expected_int = 42;
```
`int64_t`

MathMagique commented on cpp/turbodbc_numpy/Library/src/datetime_column.cpp:
```diff
@@ -25,18 +28,18 @@ namespace {
 	boost::posix_time::ptime const timestamp_epoch({1970, 1, 1}, {0, 0, 0, 0});
 
-	long timestamp_to_microseconds(char const * data_pointer)
+	intptr_t timestamp_to_microseconds(char const * data_pointer)
```
`int64_t`, no need to associate this with the size of a pointer

MathMagique commented on cpp/turbodbc_numpy/Library/src/make_numpy_array.cpp:
```diff
@@ -14,7 +14,11 @@ object make_empty_numpy_array(numpy_type const & type)
+#ifdef __GNUC__
+	                                               __extension__
+#endif
+	                                               PyArray_New(&PyArray_Type,
```
The indentation seems off here.

MathMagique commented on cpp/turbodbc_numpy/Library/src/make_numpy_array.cpp:
```diff
@@ -34,9 +38,16 @@ pybind11::object make_empty_numpy_array(std::string const & descriptor)
+#ifdef __GNUC__
+	                                               __extension__
+#endif
+	                                               PyArray_NewFromDescr(&PyArray_Type,
```
Fix indentation.

MathMagique commented on cpp/turbodbc_numpy/Library/turbodbc_numpy/datetime_column.h:
```diff
@@ -12,7 +12,7 @@ class datetime_column : public masked_column {
 	virtual ~datetime_column();
 
-	typedef std::function<long (char const * data_pointer)> converter;
+	typedef std::function<intptr_t (char const * data_pointer)> converter;
```
`int64_t`

MathMagique commented on cpp/turbodbc_python/Library/src/determine_parameter_type.cpp:
```diff
@@ -22,8 +22,8 @@ namespace {
     {
-        *reinterpret_cast<long *>(destination.data_pointer) = value.cast<long>();
-        destination.indicator = sizeof(long);
+        *reinterpret_cast<intptr_t *>(destination.data_pointer) = value.cast<intptr_t>();
```
`int64_t`

MathMagique commented on cpp/turbodbc_python/Library/src/determine_parameter_type.cpp:
```diff
@@ -83,7 +83,7 @@ namespace {
 {
     {
-        auto caster = pybind11::detail::make_caster<long>();
+        auto caster = pybind11::detail::make_caster<intptr_t>();
```
`int64_t`

MathMagique commented on cpp/turbodbc_python/Library/src/python_result_set.cpp:
```diff
@@ -23,18 +24,18 @@ namespace {
     {
         // map SQL nanosecond precision to microsecond precision
-        long const adjusted_fraction = ts.fraction / 1000;
+        intptr_t const adjusted_fraction = ts.fraction / 1000;
```
`int64_t`

MathMagique commented on cpp/turbodbc_python/Library/src/python_result_set.cpp:
```diff
@@ -23,18 +24,18 @@ namespace {
     }
 
-    object make_object(turbodbc::type_code code, char const * data_pointer, long size)
+    object make_object(turbodbc::type_code code, char const * data_pointer, intptr_t size)
```
`int64_t`

MathMagique commented on cpp/turbodbc_python/Library/src/python_result_set.cpp:
```diff
@@ -23,18 +24,18 @@ namespace {
                 return cast(*reinterpret_cast<bool const *>(data_pointer));
             case type_code::integer:
-                return cast(*reinterpret_cast<long const *>(data_pointer));
+                return cast(*reinterpret_cast<intptr_t const *>(data_pointer));
```
`int64_t`

TWAC commented:
I have tried to follow your int64_t recommendations, but I must admit I do not fully understand when which should be used, might need another pass.
Just out of curiosity, how does this work on 32-bit Linux, where I think long is 32-bit?

My formatting changes around `__extension__` was very bad, I used a macro instead.

I added back the copy-to-sdist script, it will be very hard for others to work with the repo on Windows without it. I made some progress with cmake earlier, perhaps it can go away in the future.

I also added some instructions to the readme.

TWAC commented:
Now the mac build errored out, perhaps more casts are needed? Will be hard to check since I don't have access to such a environment.
https://travis-ci.org/blue-yonder/turbodbc/jobs/214299096#L1483

MathMagique commented:
Ah, the joys of multi-platforming :-D.

To be completely honest, I am not entirely sure which types are supposed to go where exactly, because the ODBC types are often a mess. It worked with `long` for a long time, until Windows came along ;-). I think occurrences of `SQLPOINTER` that may be used as integers (sigh) should be `intptr_t`, since that represents an integer that is large enough to hold a pointer. I also think that `int64_t` is the appropriate type for the data exchange stuff, because that's what these types are bound to be. The SQLLEN seems to be something else on OSX again. I'll have to check that.

I am sure it would not work on 32 bit Linux, but the last time I worked with 32 bit Linux system was in 2006, I think. 64 bits is what I am targeting.

There's a few remaining questions for me:
* Have you successfully run the integration tests on Windows? If so, with which database/driver?
* Not coming from the Windows side, would installing the package be as simple as `pip install turbodbc` once version 1.1 is pushed to pypi?

Thanks for the changes, I'll see what I can do with the OSX build.

yaxxie commented:
> Not coming from the Windows side, would installing the package be as simple as pip install turbodbc once version 1.1 is pushed to pypi?

No, simple answer is because you're only pushing sdists at the moment, and a lot of users are unlikely to have prerequisite headers, maybe, if they have anaconda, but I never looked at this. Best way to ensure it will work like that for windows would be to push a windows bdist_wheel build, which presumably is only possible once travis has been set up with windows builds as well.

TWAC commented:
It is strange that intptr_t did not trigger the same OSX error, I would assume they are the same size.

It might be good to mention in the readme that only 64-bit is targeted.

Since the CMake build does not yet work, I have not run the tests. (And I can't get them to pass completely on Linux with PostgreSQL either).

For `pip install turbodbc` to work
- Visual C++ 2015 update 3 needs to be installed. (pybind11 has a [hard requirement](https://github.com/pybind/pybind11/blob/v2.1/include/pybind11/common.h#L54) on update 3. This unfortunately excludes the lighter [Visual C++ 2015 Build Tools](http://landinghub.visualstudio.com/visual-cpp-build-tools)). My guess is that this is somewhat common at least.
- Boost need to be installed, bootstrapped and the environment configured so that turbodbc can find it. I imagine this is very uncommon. (If not bootstrapped linking complains about libboost_date_time. When boost date handling goes away, it might be possible to do a source only build.)

In conclusion, without following the readme almost no one will get this working on Windows. If the goal is smooth install experience, a binary wheel is indeed needed.

MathMagique commented:
As it turns out, size is not really the problem. On both OSX and Linux (64 bit), the following types have the same _size_: `long`, `long long`, `int64_t`, `intptr_t`, and `SQLLEN`. However, these types basically have two internal type codes:

On Linux:

    type long: l
    type long long: x
    type int64_t: l
    type intptr_t: l
    type SQLLEN: l

On OSX:

    type long: l
    type long long: x
    type int64_t: x
    type intptr_t: l
    type SQLLEN: l

Notice how the type code for `int64_t` coincides with the type code for `SQLLEN` on Linux, but differs on OSX, though it uses the same internal representation. On Windows 64 bit I find the following:

    type long: long
    type long long: __int64
    type int64_t: __int64
    type intptr_t: __int64
    type SQLLEN: __int64

So the end result is this: `SQLLEN` matches the type code of `intptr_t`on all three platforms, probably by pure coincidence. It matches the type code of `int64_t` on two platforms only. The differing type codes essentially prevent automatic casting of pointer types since C++ will consider them separate types. In the interest of avoiding a `reinterpret_cast` (which would work), I will rework  the type in the buffer thingy to `intptr_t`.

MathMagique commented:
Thanks for the clarification concerning the status of the Windows build. Experimental seems like a good classification, then.

MathMagique commented:
@TWAC Thanks for all the hard work you put in!

MathMagique commented:
@TWAC It seems like you are not [listed as a contributor](https://github.com/blue-yonder/turbodbc/graphs/contributors) yet. I guess this may have something to do with the way your git profile on your computer and your Github profile is set up. It would be cool if you could fix that, so people know who to credit ;-).
